ares_init.c

```diff
@@ -1652,43 +1670,90 @@ static int init_by_resolv_conf(ares_channel channel)
     /* Only update search domains if they're not already specified */
     update_domains = (channel->ndomains == -1);
 
-    fp = fopen(PATH_RESOLV_CONF, "r");
-    if (fp) {
-      while ((status = ares__read_line(fp, &line, &linesize)) == ARES_SUCCESS)
-      {
-        if ((p = try_config(line, "domain", ';')) && update_domains)
-          status = config_domain(channel, p);
-        else if ((p = try_config(line, "lookup", ';')) && !channel->lookups)
-          status = config_lookup(channel, p, "bind", NULL, "file");
-        else if ((p = try_config(line, "search", ';')) && update_domains)
-          status = set_search(channel, p);
-        else if ((p = try_config(line, "nameserver", ';')) &&
-                 channel->nservers == -1)
-          status = config_nameserver(&servers, &nservers, p);
-        else if ((p = try_config(line, "sortlist", ';')) &&
-                 channel->nsort == -1)
-          status = config_sortlist(&sortlist, &nsort, p);
-        else if ((p = try_config(line, "options", ';')))
-          status = set_options(channel, p);
-        else
-          status = ARES_SUCCESS;
-        if (status != ARES_SUCCESS)
+    /* Support path for resolvconf filename set by ares_init_options */
+    if (channel->resolvconf_path) {
+      fp = fopen(channel->resolvconf_path, "r");
+      if (fp) {
+        while ((status = ares__read_line(fp, &line, &linesize)) == ARES_SUCCESS)
+        {
+          if ((p = try_config(line, "domain", ';')) && update_domains)
+            status = config_domain(channel, p);
+          else if ((p = try_config(line, "lookup", ';')) && !channel->lookups)
+            status = config_lookup(channel, p, "bind", NULL, "file");
+          else if ((p = try_config(line, "search", ';')) && update_domains)
+            status = set_search(channel, p);
+          else if ((p = try_config(line, "nameserver", ';')) &&
+                  channel->nservers == -1)
+            status = config_nameserver(&servers, &nservers, p);
+          else if ((p = try_config(line, "sortlist", ';')) &&
+                  channel->nsort == -1)
+            status = config_sortlist(&sortlist, &nsort, p);
+          else if ((p = try_config(line, "options", ';')))
+            status = set_options(channel, p);
+          else
+            status = ARES_SUCCESS;
+          if (status != ARES_SUCCESS)
+            break;
+        }
+        fclose(fp);
+      }
+      else {
+        error = ERRNO;
+        switch(error) {
+        case ENOENT:
+        case ESRCH:
+          status = ARES_EOF;
           break;
+        default:
+          DEBUGF(fprintf(stderr, "fopen() failed with error: %d %s\n",
+                        error, strerror(error)));
+          DEBUGF(fprintf(stderr, "Error opening file: %s\n", PATH_RESOLV_CONF));
+          status = ARES_EFILE;
+        }
       }
-      fclose(fp);
     }
```

Comment:
Too much duplication here of logic.  If someone explicitly configured a resolveconf_path, I don't see any reason to use the default PATH_RESOLV_CONF.  So really we should just do this once.  I'd think it makes more sense to error out anyhow.